### PR TITLE
[PAN-2333] Use only fully validated peers for fast sync pivot selection

### PR DESCRIPTION
**Pull request by mbaxter**
## PR description
Expose `EthPeer` validation state so that the `Synchronizer` can choose peers based on whether or not they have been fully validated.  This allows us to use only fully validated peers when choosing a pivot block.

### CODE REVIEW

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java [hunk 1]
```diff
@@ -34,39 +35,57 @@
...
+    verify(validator, never()).validatePeer(eq(ethProtocolManager.ethContext()), eq(peer));
     verify(runner, never()).disconnectPeer(eq(peer));
     verify(runner, times(2)).scheduleNextCheck(eq(peer));
+
+    assertThat(peer.isFullyValidated()).isFalse();
   }
 
   @Test
   public void checkPeer_doesNotScheduleFutureCheckWhenPeerNotReadyAndDisconnected() {
+    PeerValidator validator = mock(PeerValidator.class);
```
Comment (abdelhamidbakhta) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java]: (suggestion) make the variable `final`

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java [hunk 2]
```diff
@@ -34,39 +35,57 @@
...
+
     EthProtocolManager ethProtocolManager = EthProtocolManagerTestUtil.create();
     EthProtocolManagerTestUtil.disableEthSchedulerAutoRun(ethProtocolManager);
-    EthPeer peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager).getEthPeer();
+    EthPeer peer =
```
Comment (abdelhamidbakhta) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java]: (suggestion) make the variable `final`

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java [hunk 3]
```diff
@@ -75,47 +94,71 @@ public void checkPeer_doesNotScheduleFutureCheckWhenPeerNotReadyAndDisconnected(
...
     verify(runner, times(1)).checkPeer(eq(peer));
-    verify(validator, never()).validatePeer(eq(peer));
+    verify(validator, never()).validatePeer(eq(ethProtocolManager.ethContext()), eq(peer));
     verify(runner, never()).disconnectPeer(eq(peer));
     verify(runner, times(0)).scheduleNextCheck(eq(peer));
   }
 
   @Test
   public void checkPeer_handlesInvalidPeer() {
+    PeerValidator validator = mock(PeerValidator.class);
```
Comment (abdelhamidbakhta) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java]: (suggestion) make the variable `final`

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java [hunk 4]
```diff
@@ -75,47 +94,71 @@ public void checkPeer_doesNotScheduleFutureCheckWhenPeerNotReadyAndDisconnected(
...
+
     EthProtocolManager ethProtocolManager = EthProtocolManagerTestUtil.create();
     EthProtocolManagerTestUtil.disableEthSchedulerAutoRun(ethProtocolManager);
-    EthPeer peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager).getEthPeer();
+    EthPeer peer =
```
Comment (abdelhamidbakhta) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java]: (suggestion) make the variable `final`

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java [hunk 5]
```diff
@@ -75,47 +94,71 @@ public void checkPeer_doesNotScheduleFutureCheckWhenPeerNotReadyAndDisconnected(
...
+    verify(validator, times(1)).validatePeer(eq(ethProtocolManager.ethContext()), eq(peer));
     verify(runner, times(1)).disconnectPeer(eq(peer));
     verify(runner, never()).scheduleNextCheck(eq(peer));
+
+    assertThat(peer.isFullyValidated()).isFalse();
   }
 
   @Test
   public void checkPeer_handlesValidPeer() {
+    PeerValidator validator = mock(PeerValidator.class);
```
Comment (abdelhamidbakhta) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java]: (suggestion) make the variable `final`

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java [hunk 6]
```diff
@@ -75,47 +94,71 @@ public void checkPeer_doesNotScheduleFutureCheckWhenPeerNotReadyAndDisconnected(
...
+
     EthProtocolManager ethProtocolManager = EthProtocolManagerTestUtil.create();
     EthProtocolManagerTestUtil.disableEthSchedulerAutoRun(ethProtocolManager);
-    EthPeer peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager).getEthPeer();
+    EthPeer peer =
```
Comment (abdelhamidbakhta) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java]: (suggestion) make the variable `final`

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthProtocolManager.java [hunk 7]
```diff
@@ -90,6 +95,11 @@ public EthProtocolManager(
 
     this.blockBroadcaster = new BlockBroadcaster(ethContext);
 
+    // Run validators
+    for (final PeerValidator peerValidator : this.peerValidators) {
```
[the remark this answers is not in the document]
Comment (mbaxter) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthProtocolManager.java]: The `runValidator` method just sets up a listener, so I think we should be fine with a plain loop here.

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java (hunk 1 above)
Comment (mbaxter) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/peervalidation/PeerValidatorRunnerTest.java]: 👍 updated!

**Pull request merged**